src/StreamMaster.cpp

```diff
@@ -0,0 +1,203 @@
+#include "StreamMaster.h"
+#include "Streamer.h"
+
+FileWriter::StreamMaster::StreamMaster(
+    const std::string &Broker, std::unique_ptr<FileWriterTask> FileWriterTask,
+    const MainOpt &Options,
+    std::unique_ptr<StreamerBaseFactory> StreamerFactory)
+    : Demuxers(FileWriterTask->demuxers()),
+      ProducerTopic(FileWriterTask->getStatusProducer()),
+      WriterTask(std::move(FileWriterTask)),
+      TopicWriteDuration{Options.topic_write_duration},
+      ServiceId{Options.service_id} {
+
+  for (auto &Demux : Demuxers) {
+    try {
+      Streamers.emplace(std::piecewise_construct,
+                        std::forward_as_tuple(Demux.topic()),
+                        std::forward_as_tuple(StreamerFactory->create(
+                            Broker, Demux, Options.StreamerConfiguration)));
+    } catch (std::exception &E) {
+      RunStatus = StreamMasterError::STREAMER_ERROR();
+      LOG(Sev::Critical, "{}", E.what());
+      logEvent(ProducerTopic, StatusCode::Error, ServiceId,
+               WriterTask->job_id(), E.what());
+    }
+  }
+  NumStreamers = Streamers.size();
+  ReportMessageDelay =
+      std::chrono::milliseconds{Options.status_master_interval};
+}
+
+FileWriter::StreamMaster::~StreamMaster() {
+  Stop = true;
+  if (WriteThread.joinable()) {
+    WriteThread.join();
+  }
+  if (ReportThread.joinable()) {
+    ReportThread.join();
+  }
+  LOG(Sev::Info, "Stop StreamMaster for file with id : {}, ready to be removed",
+      getJobId());
+}
+
+bool FileWriter::StreamMaster::setStopTime(
+    const std::chrono::milliseconds &StopTime) {
+  for (auto &s : Streamers) {
+    try {
+      //        s.second->getOptions().StopTimestamp = StopTime;
+    } catch (std::exception &E) {
+      LOG(Sev::Error, "Unable to set stop time in topic {}: {}", s.first,
+          E.what());
+    }
+  }
+  return true;
+}
+
+bool FileWriter::StreamMaster::start() {
+  runReport(ReportMessageDelay);
+  if (NumStreamers == 0) {
+    Stop = true;
+    stopImplemented();
+    return WriteThread.joinable();
+  }
+  LOG(Sev::Info, "StreamMaster: start");
+  Stop = false;
+
+  if (!WriteThread.joinable()) {
```

Comment:
Should this test for a not-yet-started `WriteThread`? It can also be that `WriteThread.joinable() == false` in the case that the thread was already run and joined before.